File: python/supersendtx/http.py
from __future__ import annotations

import json
import urllib.error
import urllib.parse
import urllib.request
from typing import Any

from supersendtx.errors import SuperSendTXError


class HttpClient:
    def __init__(self, api_key: str, *, base_url: str = "https://api.supersendtx.com") -> None:
        if not api_key.startswith("stx_"):
            raise ValueError("SuperSend TX API key must start with stx_")
        self.api_key = api_key
        self.base_url = base_url.rstrip("/")

    def request(
        self,
        method: str,
        path: str,
        *,
        body: Any = None,
        headers: dict[str, str] | None = None,
    ) -> Any:
        url = f"{self.base_url}{path}"
        payload = None
        req_headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
            **(headers or {}),
        }
        if body is not None:
            payload = json.dumps(body).encode("utf-8")
        request = urllib.request.Request(url, data=payload, headers=req_headers, method=method)
        try:
            with urllib.request.urlopen(request) as response:
                raw = response.read().decode("utf-8")
                if not raw:
                    return {}
                return json.loads(raw)
        except urllib.error.HTTPError as error:
            raw = error.read().decode("utf-8")
            try:
                parsed = json.loads(raw)
            except json.JSONDecodeError:
                parsed = {}
            raise SuperSendTXError.from_response(error.code, parsed) from error

    @staticmethod
    def query(params: dict[str, Any]) -> str:
        filtered = {key: value for key, value in params.items() if value is not None}
        if not filtered:
            return ""
        return "?" + urllib.parse.urlencode(filtered)

File: ruby/lib/supersendtx/http.rb
# frozen_string_literal: true

require "json"
require "net/http"
require "uri"

module SuperSendTX
  class HttpClient
    DEFAULT_API_BASE_URL = "https://api.supersendtx.com"

    def initialize(api_key, base_url: DEFAULT_API_BASE_URL, transport: nil)
      raise ArgumentError, "SuperSend TX API key must start with stx_ or rnl_" unless api_key.start_with?("stx_", "rnl_")

      @api_key = api_key
      @base_url = base_url.chomp("/")
      @transport = transport
    end

    def request(method, path, body: nil, headers: {})
      request_headers = {
        "Authorization" => "Bearer #{@api_key}",
        "Content-Type" => "application/json"
      }.merge(headers)

      return @transport.call(method, path, body, request_headers) if @transport

      uri = URI.parse("#{@base_url}#{path}")
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = uri.scheme == "https"

      request_class = Net::HTTP.const_get(method.capitalize)
      request = request_class.new(uri.request_uri)
      request_headers.each { |key, value| request[key] = value }
      request.body = JSON.generate(body) if body

      response = http.request(request)
      parsed = response.body.to_s.empty? ? {} : JSON.parse(response.body)

      raise Error.from_response(response.code.to_i, parsed) if response.code.to_i >= 400

      parsed
    end

    def query(params)
      filtered = params.compact
      return "" if filtered.empty?

      "?#{URI.encode_www_form(filtered)}"
    end
  end
end
